asynction/types.py

```diff
@@ -13,11 +15,223 @@
+    """
+    https://www.asyncapi.com/docs/specifications/v2.2.0#oauthFlowsObject
+    """
+
+    IMPLICIT = "implicit"
+    PASSWORD = "password"
+    CLIENT_CREDENTIALS = "clientCredentials"
+    AUTHORIZATION_CODE = "authorizationCode"
+
+
+@dataclass
+class OAuth2Flow:
+    """
+    https://www.asyncapi.com/docs/specifications/v2.2.0#oauthFlowObject
+    """
+
+    scopes: Mapping[str, str]
+    authorization_url: Optional[str] = None
+    token_url: Optional[str] = None
+    refresh_url: Optional[str] = None
+
+    @staticmethod
+    def forge(
+        type_: Type["OAuth2Flow"], data: JSONMapping, forge: Forge
+    ) -> "OAuth2Flow":
+        return type_(
+            scopes=forge(type_.__annotations__["scopes"], data.get("scopes")),
+            authorization_url=forge(
+                type_.__annotations__["authorization_url"], data.get("authorizationUrl")
+            ),
+            token_url=forge(type_.__annotations__["token_url"], data.get("tokenUrl")),
+            refresh_url=forge(
+                type_.__annotations__["refresh_url"], data.get("refreshUrl")
+            ),
+        )
+
+
+register_forge(OAuth2Flow, OAuth2Flow.forge)
+
+
+@dataclass
+class OAuth2Flows:
+    implicit: Optional[OAuth2Flow] = None
+    password: Optional[OAuth2Flow] = None
+    client_credentials: Optional[OAuth2Flow] = None
+    authorization_code: Optional[OAuth2Flow] = None
+
+    @staticmethod
+    def forge(
+        type_: Type["OAuth2Flows"], data: JSONMapping, forge: Forge
+    ) -> "OAuth2Flows":
+        return type_(
+            implicit=forge(type_.__annotations__["implicit"], data.get("implicit")),
+            password=forge(type_.__annotations__["password"], data.get("password")),
+            client_credentials=forge(
+                type_.__annotations__["client_credentials"],
+                data.get("clientCredentials"),
+            ),
+            authorization_code=forge(
+                type_.__annotations__["authorization_code"],
+                data.get("authorizationCode"),
+            ),
+        )
+
+    def __post_init__(self):
+        if self.implicit is not None and self.implicit.authorization_url is None:
+            raise ValueError("Implicit OAuth flow is missing Authorization URL")
+        elif self.password is not None and self.password.token_url is None:
+            raise ValueError("Password OAuth flow is missing Token URL")
+        elif (
+            self.client_credentials is not None
+            and self.client_credentials.token_url is None
+        ):
+            raise ValueError("Client Credentials OAuth flow is missing Token URL")
+        elif (
+            self.authorization_code is not None
+            and self.authorization_code.token_url is None
+        ):
+            raise ValueError("Authorization code OAuth flow is missing Token URL")
+
+    def supported_scopes(self) -> Iterator[str]:
+        # note Cannot lru_cache this
+        # TypeError: unhashable type: 'OAuth2Flows'
```

Comment:
FYI: you can make a dataclass hashable by passing the `frozen` argument in the dataclass decorator: https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass

Not important though.